File: RoadFlow.Identity.Core/Domains/Password/PasswordValidator.cs
using System.Text.RegularExpressions;
using FluentValidation;

namespace RoadFlow.Identity.Core.Domains.Password;

public class PasswordValidator : AbstractValidator<string>
{
    /*
    * minimum 8 characters
    * maximum 32 characters
    * at least one uppercase letter
    * one lowercase letter
    * one digit
    * one special character
    */
    private readonly Regex _regex = new("^(?=.*[a-z])(?=.*[A-Z])(?=.*\\d)(?=.*[@$!%*?&])[A-Za-z\\d@$!%*?&]{8,32}$");
    
    public PasswordValidator()
    {
        RuleFor(x => x)
            .NotEmpty()
            .Matches(_regex);
    }
}

File: road-flow-client/src/shared/validators/password-validator.ts
import * as yup from 'yup';

const passwordValidator = yup
    .string()
    .required('Password is required')
    .min(8, 'Minimum length of password is 8 characters')
    .max(32, 'Maximum length of password is 32 characters')
    .matches(/[A-Z]+/, 'Password must contains at a least one uppercase letter')
    .matches(/[a-z]+/, 'Password must contains at a least one lowercase letter')
    .matches(/\d+/, 'Password must contains at a least one digit')
    .matches(/[@$!%*?&]+/, 'Password must contains at a least one special symbol');

export default passwordValidator;